src/Mesh/Metrics.jl

```diff
@@ -291,6 +292,7 @@ function computemetric!(x1, x2, x3, J, ξ1x1, ξ2x1, ξ3x1, ξ1x2, ξ2x2, ξ3x2,
         x3ξ2 += D[j, n] * x3[i, n, k, e]
         x3ξ3 += D[k, n] * x3[i, j, n, e]
       end
+      JH[i, j, k, e] = x1ξ1 * x2ξ2 - x2ξ1 * x1ξ2
```

Comment:
I see, so we would need the face normals aswell?

Reply:
I'm not sure I follow, the surface metrics are naturally calculated completely from the normals. What in the surface metrics should I be trying to imitate in order to be able to capture the warping, I was under the impression the warping was taken into account here : https://github.com/climate-machine/CLIMA/blob/3e72b701ae6192309ef51f2fcc65ecde11663631/src/Mesh/Grids.jl#L271

Reply:
That gives you the points, but to compute integrals you need relationships between the derivatives. To do this you need something like what's done for the [surface Jacobians](https://github.com/climate-machine/CLIMA/blob/c10b76acadc62e89fed459420dbd1bcd5011a9fc/src/Mesh/Metrics.jl#L395-L397)

Reply:
I'm confused, do you mean having something along the lines of hypot(x1ξ1 * (x2ξ2 * x3ξ3 - x3ξ2 * x2ξ3) ,x2ξ1 * (x3ξ2 * x1ξ3 - x1ξ2 * x3ξ3), x3ξ1 * (x1ξ2 * x2ξ3 - x2ξ2 * x1ξ3))

Reply:
I don't think that's quite right. I think you want something along the lines of `hypot(J * ξ3x1, J * ξ3x2, J * ξ3x3)' but I wouldn't know for sure unless I worked it out on paper. This is the surface Jacobian calculation I highlight on faces 5 & 6.

If you add a reasonable tests and it passes, then this would show that what you have is correct.

Reply:
Oh alright I see what you mean now.

Reply:
`JcV` is for a line integral. For a surface integral you need to consider a parameterized surface integral; see for instance this [wikipedia article](https://en.wikipedia.org/wiki/Surface_integral#Surface_integrals_of_scalar_fields).

Using the notation from the article, you need to compute:
<img src="https://latex.codecogs.com/gif.latex?\iint_Sf(\vec{x})&space;dS&space;=&space;\int_{-1}^{1}\int_{-1}^{1}f\left(\vec{x}\left(\vec{\xi}\right)\right)\left\|\frac{\partial&space;\vec{x}}{\partial\xi_1}\times\frac{\partial&space;\vec{x}}{\partial\xi_2}\right\|d\xi_1d\xi_2"/>

Considering the cross product term gives
<img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;\vec{x}}{\xi_1}\times\frac{\partial&space;\vec{x}}{\xi_2}&space;=&space;\begin{bmatrix}&space;\frac{\partial&space;x_2}{\partial\xi_1}&space;\frac{\partial&space;x_3}{\partial\xi_2}-\frac{\partial&space;x_3}{\partial&space;\xi_1}&space;\frac{\partial&space;x_2}{\partial\xi_2}\\&space;\frac{\partial&space;x_3}{\partial\xi_1}&space;\frac{\partial&space;x_1}{\partial\xi_2}-\frac{\partial&space;x_1}{\partial&space;\xi_1}&space;\frac{\partial&space;x_3}{\partial\xi_2}\\&space;\frac{\partial&space;x_1}{\partial\xi_1}&space;\frac{\partial&space;x_2}{\partial\xi_2}-\frac{\partial&space;x_2}{\partial&space;\xi_1}&space;\frac{\partial&space;x_1}{\partial\xi_2}\\&space;\end{bmatrix}&space;=&space;\begin{bmatrix}&space;J\frac{\partial&space;\xi_1}{\partial&space;x_3}\\&space;J\frac{\partial&space;\xi_2}{\partial&space;x_3}\\&space;J\frac{\partial&space;\xi_3}{\partial&space;x_3}\\&space;\end{bmatrix}&space;=&space;J\frac{\partial&space;\vec{\xi}}{\partial&space;x_3}"/>
where the last equality follows from the standard metric identities; see for instance [Kopriva (2006) equation (5)](https://doi.org/10.1007/s10915-005-9070-8).

Thus we have that
<img src="https://latex.codecogs.com/gif.latex?\left\|\frac{\partial&space;\vec{x}}{\xi_1}\times\frac{\partial&space;\vec{x}}{\xi_2}\right\|&space;=&space;\text{hypot}\left(J\frac{\partial&space;\xi_1}{\partial&space;x_3},&space;J\frac{\partial&space;\xi_2}{\partial&space;x_3},&space;J\frac{\partial&space;\xi_3}{\partial&space;x_3}\right)"/>

Reply:
I went with the following done on Grids.jl. Is there any reason this should not work? before I go ahead try to set up a minimum working case.

MH = kron(1, ntuple(j->ω, dim-1)...)
map!(JH, J, ξ1x3, ξ2x3, ξ3x3
        ) do J, ξ1x3, ξ2x3, ξ3x3
      hypot(J*ξ1x3 ,J*ξ2x3, J*ξ3x3)
    end
    MHJH .= JH .* MH

Reply:
@jkozdon the latest push has a test case based on warp fun. The integrals for it should always be =0.5 but it seems the result is always a bit off, I'd appreciate your input.

Reply:
Hi @jkozdon I think I may be off because I'm still using "MH = kron(1, ntuple(j->ω, dim)...)" and as such might have the wrong weights, however "MH = kron(1, ntuple(j->ω, dim-1)...)" doesn't work because the array sizes then don't seem to match is this what the syntax should be?

Reply:
I think you need something like: `MH = kron(ones(FT, Nq), ntuple(j->ω, dim-1)...)` since you want to replicate the 2D reference matrix `Nq` times (one for each 2D slab of degrees of freedom).

Reply:
Alright I'll try that, thanks

Reply:
This [comment](https://github.com/climate-machine/CLIMA/pull/504#discussion_r340769742) had a typo.

The cross product terms should be
<img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;\vec{x}}{\partial\xi_1}\times\frac{\partial&space;\vec{x}}{\partial\xi_2}&space;=&space;\begin{bmatrix}&space;J\frac{\partial&space;\xi_3}{\partial&space;x_1}\\&space;J\frac{\partial&space;\xi_3}{\partial&space;x_2}\\&space;J\frac{\partial&space;\xi_3}{\partial&space;x_3}\\&space;\end{bmatrix}&space;=&space;J\frac{\partial&space;\xi_3}{\partial&space;\vec{x}}"/>
(see [Kopriva (2006) equation (5)](https://doi.org/10.1007/s10915-005-9070-8)) which then gives
<img src="https://latex.codecogs.com/gif.latex?\left\|\frac{\partial&space;\vec{x}}{\partial\xi_1}\times\frac{\partial&space;\vec{x}}{\partial\xi_2}\right\|&space;=&space;\text{hypot}\left(J\frac{\partial&space;\xi_3}{\partial&space;x_1},&space;J\frac{\partial&space;\xi_3}{\partial&space;x_2},&space;J\frac{\partial&space;\xi_3}{\partial&space;x_3}\right)"/>